README.md
# i2c-adt7420

[![npm version](https://img.shields.io/npm/v/i2c-adt7420.svg)](https://www.npmjs.com/package/i2c-adt7420)
[![npm license](https://img.shields.io/npm/l/i2c-adt7420.svg)](https://www.npmjs.com/package/i2c-adt7420)
[![tests](https://github.com/tvdstaaij/node-i2c-adt7420/actions/workflows/test.yml/badge.svg)](https://github.com/tvdstaaij/node-i2c-adt7420/actions/workflows/test.yml)

Node.js driver for the acurate (±0.25°C) yet affordable [ADT7420][adt7420-prod] digital temperature sensor.
Should also work with the less accurate [ADT7410][adt7410-prod].

This module was tested with Raspberry Pi 3B and 4B, wired to an [EVAL-ADT7420-PMDZ][adt7420-eval] breakout board.
Please do note that the power LED on the EVAL-ADT7420-PMDZ board significantly influences the temperature reading!
The LED or its series resistor should be desoldered if you need accurate temperature.

Supported features:

* 13 and 16 bit resolution
* Continuous and one sample per second modes
* Alternate I2C addresses

## Installation

Available from npm: `npm install i2c-adt7420`

Requires Node.js 14 or higher.

## Example: ADT7420 temperature server

```javascript
const express = require('express');
const {ADT7420} = require('i2c-adt7420');

const config = {
  pollingInterval: 1000,
  httpPort: 7420,
  i2cBusNumber: 1
};

const app = express();
let adt7420 = null;
let lastUpdateTime = null;
let temperature = null;

app.get('/', (req, res) => {
  res.setHeader('Content-Type', 'application/json');
  res.send(JSON.stringify({
    temperature: {
      raw: temperature.raw,
      celsius: temperature.celsius,
      kelvin: temperature.kelvin
    },
    time_updated: lastUpdateTime
  }, null, 2));
})

ADT7420.open({
  i2cBusNumber: config.i2cBusNumber
})
  .then((instance) => {
    adt7420 = instance;
    return adt7420.configure({
      resolution: 16,
      lowPowerMode: (config.pollingInterval >= 1000)
    });
  })
  .then(() => {
    console.log('Configured ADT7420');
    return updateTemperature();
  })
  .then(() => {
    console.log(`Read temperature of ${temperature.celsius}C`);
    setInterval(updateTemperature, config.pollingInterval);
    app.listen(config.httpPort, () => {
      console.log(`Serving temperature data on port ${config.httpPort}`);
    });
  })
  .catch((err) => {
    console.error(err);
    process.exit(3);
  });

function updateTemperature() {
  return adt7420.readTemperature()
    .then((result) => {
      lastUpdateTime = new Date();
      temperature = result;
    })
    .catch((err) => {
      console.error(err);
      process.exit(3);
    });
}
```

## API

### Class `ADT7420`

`ADT7420.open([options]) -> Promise`

Factory method to asynchronously create an instance of the class.

Arguments:

* `options` object of options:
    * `i2cBusNumber` Optional integer specifying the system's I2C device number of the bus the chip is attached to.
      If you use a Raspberry Pi, this should typically be `1` for a Raspberry Pi 2+ and `0` for the original model.
      Make sure the operating system supports I2C and that it is enabled. Default: 0.
    * `i2cAddress` Optional, one of `ADT7420.I2C_ADDRESS_0x48` (default), `ADT7420.I2C_ADDRESS_0x49`, `ADT7420.I2C_ADDRESS_0x4A`, `ADT7420.I2C_ADDRESS_0x4B`.
      This only needs to be specified if you have changed the default address of the chip by physically connecting its address configuration pins.

Returns: promise resolving to an ADT7420 class instance when succesful.

`ADT7420#configure([options]) -> Promise`

Changes the chip's configuration register. Configuring is recommended but not mandatory.

Arguments:

* `options` object of options:
    * `resolution` Optional integer specifying the sample resolution in bits (either 13 or 16).
      The chip's reset value is 13 bits.
    * `lowPowerMode` Optional boolean, false for continuous sample mode, true for one sample per second mode. 
      The chip's reset value is false (continuous mode). See datasheet for details.

Returns: promise resolving to undefined when succesful.

`ADT7420#readTemperature() -> Promise`

Reads the chip's temperature register.

Returns: promise resolving to an `ADT7420Temperature` instance when succesful.

### Class `ADT7420Temperature`

`ADT7420Temperature#raw : Number`

Raw temperature reading as a 16-bit two's complement integer.

`ADT7420Temperature#celsius : Number`

Temperature reading converted to degrees Celsius.

`ADT7420Temperature#kelvin : Number`

Temperature reading converted to Kelvin.

[adt7420-prod]: http://www.analog.com/en/products/analog-to-digital-converters/integrated-special-purpose-converters/digital-temperature-sensors/adt7420.html
[adt7410-prod]: http://www.analog.com/en/products/analog-to-digital-converters/integrated-special-purpose-converters/digital-temperature-sensors/adt7410.html
[adt7420-eval]: http://www.analog.com/en/design-center/evaluation-hardware-and-software/evaluation-boards-kits/eval-adt7420-pmdz.html


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2022-08-13

### Changed
- Update all dependencies to latest versions.
- Change CI from Travis to Github Actions.

### Removed
- Support for Node.js versions older than 14.


ci/test.sh
#!/bin/bash
set -Eeuo pipefail

npm ci
npm test


test/index.js
'use strict';

require('./adt7420-stubbed');


test/adt7420-stubbed.js
'use strict';

const should = require('chai').should();
const assert = require('chai').assert;
const ADT7420 = require('../lib/adt7420');
const ADT7420Stub = require('./lib/adt7420-stub');

describe('ADT7420 (stubbed)', function() {

  it('should read temperature with 13-bit resolution', function() {
    const adt7420Stub = new ADT7420Stub();
    const adt7420 = new ADT7420(adt7420Stub, ADT7420.I2C_ADDRESS_DEFAULT);
    const roomTemperature = 20;
    const fractionalTemperature = 20.4242424242;
    adt7420Stub.setTemperature(
      ADT7420.Temperature.fromCelsius(roomTemperature, 13));
    return adt7420.readTemperature()
      .then((temperatureReading) => {
        temperatureReading.celsius.should.equal(roomTemperature);
        adt7420Stub.setTemperature(
          ADT7420.Temperature.fromCelsius(fractionalTemperature, 13));
        return adt7420.readTemperature();
      })
      .then((temperatureReading) => {
        temperatureReading.celsius.should.equal(20.4375);
      });
  });

  it('should read temperature with 16-bit resolution', function() {
    const adt7420Stub = new ADT7420Stub();
    const adt7420 = new ADT7420(adt7420Stub, ADT7420.I2C_ADDRESS_DEFAULT);
    const roomTemperature = 20;
    const fractionalTemperature = 20.4242424242;
    adt7420Stub.setTemperature(
      ADT7420.Temperature.fromCelsius(roomTemperature, 16));
    return adt7420.configure({resolution: 16})
      .then(() => adt7420.readTemperature())
      .then((temperatureReading) => {
        temperatureReading.celsius.should.equal(roomTemperature);
        adt7420Stub.setTemperature(
          ADT7420.Temperature.fromCelsius(fractionalTemperature, 16));
        return adt7420.readTemperature();
      })
      .then((temperatureReading) => {
        temperatureReading.celsius.should.equal(20.421875);
      });
  });

  it('should read negative temperature', function() {
    const adt7420Stub = new ADT7420Stub();
    const adt7420 = new ADT7420(adt7420Stub, ADT7420.I2C_ADDRESS_DEFAULT);
    const siberianTemperature = -20;
    adt7420Stub.setTemperature(
      ADT7420.Temperature.fromCelsius(siberianTemperature, 16));
    return adt7420.configure({resolution: 16})
      .then(() => adt7420.readTemperature())
      .then((temperatureReading) => {
        temperatureReading.celsius.should.equal(siberianTemperature);
      });
  });

  it('should work with an alternate I2C address', function() {
    const adt7420Stub = new ADT7420Stub(0x4A);
    const adt7420 = new ADT7420(adt7420Stub, ADT7420.I2C_ADDRESS_0x4A);
    const roomTemperature = 20;
    adt7420Stub.setTemperature(
      ADT7420.Temperature.fromCelsius(roomTemperature, 13));
    return adt7420.readTemperature()
      .then((temperatureReading) => {
        temperatureReading.celsius.should.equal(roomTemperature);
      });
  });

  it('should configure continuous and low power mode', function() {
    const adt7420Stub = new ADT7420Stub();
    const adt7420 = new ADT7420(adt7420Stub, ADT7420.I2C_ADDRESS_DEFAULT);
    return adt7420.configure({lowPowerMode: false})
      .then(() => {
        adt7420Stub.getConfiguredMode()
          .should.equal(ADT7420Stub.MODE_CONTINUOUS);
        return adt7420.configure({lowPowerMode: true});
      })
      .then(() => {
        adt7420Stub.getConfiguredMode().should.equal(ADT7420Stub.MODE_1SPS);
      });
  });

  it('should propagate errors through the promise chain', function() {
    const adt7420Stub = new ADT7420Stub(0x4A); // Mismatched address
    const adt7420 = new ADT7420(adt7420Stub, ADT7420.I2C_ADDRESS_DEFAULT);
    return adt7420.readTemperature()
      .then(() => {
        throw new Error('Should not succeed');
      }, (err) => {
        should.exist(err);
      });
  });

});



lib/adt7420-temperature.js
'use strict';

class ADT7420Temperature {
  constructor(raw) {
    this._raw = raw;
  }
  static fromTemperatureRegister(registerValue, resolution) {
    switch (resolution) {
      case 13:
        return new ADT7420Temperature(registerValue & (~0x07));
      case 16:
        return new ADT7420Temperature(registerValue);
      default:
        throw new Error('Invalid resolution');
    }
  }
  static fromCelsius(degrees, resolution) {
    let raw;
    switch (resolution) {
      case 13:
        raw = Math.round(degrees * 16) * 8;
        break;
      case 16:
        raw = Math.round(degrees * 128);
        break;
      default:
        throw new Error('Invalid resolution');
    }
    if (raw >= 0) {
      // Positive temperature
      return new ADT7420Temperature(raw);
    } else {
      // Negative temperature; convert to 16-bit two's complement
      return new ADT7420Temperature(raw + 0x10000);
    }
  }
  get raw() {
    return this._raw;
  }
  get celsius() {
    if (this._raw < 0x8000) {
      // Positive temperature
      return this._raw / 128;
    } else {
      // Negative temperature; convert from 16-bit two's complement
      return (this._raw - 0x10000) / 128;
    }
  }
  get kelvin() {
    return this.celsius + 273.15;
  }
}

module.exports = ADT7420Temperature;


lib/adt7420.js
'use strict';

const i2c = require('i2c-bus');
const ADT7420Temperature = require('./adt7420-temperature');

const REG_CONFIG_ADDRESS = 0x03;
const REG_CONFIG_RESOLUTION_13 = 0x00;
const REG_CONFIG_RESOLUTION_16 = 0x80;
const REG_CONFIG_1SPS = 0x40;

class ADT7420 {
  static get I2C_ADDRESS_DEFAULT() { return 0x48; }
  static get I2C_ADDRESS_0x48() { return 0x48; }
  static get I2C_ADDRESS_0x49() { return 0x49; }
  static get I2C_ADDRESS_0x4A() { return 0x4A; }
  static get I2C_ADDRESS_0x4B() { return 0x4B; }

  constructor(i2cBus, i2cAddress) {
    this._i2cBus = i2cBus;
    this._i2cAddress = i2cAddress;
    this._chipConfig = {
      // Defaults after sensor reset
      resolution: 13,
      lowPowerMode: false
    };
  }

  static open(options) {
    options = options || {};
    const i2cAddress = Number.isInteger(options.i2cAddress) ?
      options.i2cAddress : ADT7420.I2C_ADDRESS_DEFAULT;
    const i2cBusNumber = Number.isInteger(options.i2cBusNumber) ?
      options.i2cBusNumber : 0;
    return new Promise((resolve, reject) => {
      const i2cBus = i2c.open(i2cBusNumber, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve(i2cBus);
        }
      });

    })
      .then((i2cBus) => {
        return new ADT7420(i2cBus, i2cAddress);
      });
  }

  configure(options) {
    options = Object.assign({}, options);
    if (options.resolution === 13) {
      options.resolution = REG_CONFIG_RESOLUTION_13;
    } else if (options.resolution === 16) {
      options.resolution = REG_CONFIG_RESOLUTION_16;
    } else if (options.resolution !== undefined) {
      throw new Error('resolution must be 13 or 16');
    } else {
      options.resolution = this._chipConfig.resolution;
    }
    if (options.lowPowerMode === undefined) {
      options.lowPowerMode = this._chipConfig.lowPowerMode;
    } else if (options.lowPowerMode !== true &&
               options.lowPowerMode !== false) {
      throw new Error('lowPowerMode must be a boolean');
    }

    let configRegValue = options.resolution;
    if (options.lowPowerMode) {
      configRegValue |= REG_CONFIG_1SPS;
    }
    return new Promise((resolve, reject) => {
      this._i2cBus.writeByte(
        this._i2cAddress, REG_CONFIG_ADDRESS, configRegValue, (err) => {
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        }
      );
    })
      .then(() => {
        this._chipConfig = options;
      });
  }

  readTemperature() {
    const buf = Buffer.alloc(2);
    return new Promise((resolve, reject) => {
      this._i2cBus.i2cRead(this._i2cAddress, 2, buf, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      })
    })
      .then(() => {
        let rawTemperature = (buf[0] << 8) | buf[1];
        if (this._chipConfig.resolution === ADT7420.RESOLUTION_13_BIT) {
          rawTemperature &= (~0x07);
        }
        return new ADT7420Temperature(rawTemperature,
                                      this._chipConfig.resolution);
      });
  }
}

ADT7420.Temperature = ADT7420Temperature;

module.exports = ADT7420;
